vmm/src/controller.rs

```diff
@@ -0,0 +1,1006 @@
+            let epoll_config = builder
+                .inner_mut()
+                .epoll_context_mut()
+                .allocate_tokens_for_virtio_device(TYPE_NET, &cfg.iface_id, NET_EVENTS_COUNT);
+
+            let allow_mmds_requests = cfg.allow_mmds_requests();
+
+            let rx_rate_limiter = cfg
+                .rx_rate_limiter
+                .map(vmm_config::RateLimiterConfig::into_rate_limiter)
+                .transpose()
+                .map_err(CreateRateLimiter)?;
+
+            let tx_rate_limiter = cfg
+                .tx_rate_limiter
+                .map(vmm_config::RateLimiterConfig::into_rate_limiter)
+                .transpose()
+                .map_err(CreateRateLimiter)?;
+
+            let tap = cfg.open_tap().map_err(|_| NetDeviceNotConfigured)?;
+
+            let net_box = Box::new(
+                devices::virtio::Net::new_with_tap(
+                    tap,
+                    cfg.guest_mac(),
+                    epoll_config,
+                    rx_rate_limiter,
+                    tx_rate_limiter,
+                    allow_mmds_requests,
+                )
+                .map_err(CreateNetDevice)?,
+            );
+
+            builder.attach_device(
+                cfg.iface_id.clone(),
+                MmioDevice::new(builder.inner().guest_memory().clone(), net_box).map_err(|e| {
+                    RegisterMMIODevice(super::device_manager::mmio::Error::CreateMmioDevice(e))
+                })?,
+            )?;
+        }
+
+        Ok(())
+    }
+
+    fn attach_vsock_device(
+        &mut self,
+        builder: &mut VmmBuilder,
+    ) -> result::Result<(), StartMicrovmError> {
+        if let Some(cfg) = &self.device_configs.vsock {
+            let backend = devices::virtio::vsock::VsockUnixBackend::new(
+                u64::from(cfg.guest_cid),
+                cfg.uds_path.clone(),
+            )
+            .map_err(StartMicrovmError::CreateVsockBackend)?;
+
+            let epoll_config = builder
+                .inner_mut()
+                .epoll_context_mut()
+                .allocate_tokens_for_virtio_device(TYPE_VSOCK, &cfg.vsock_id, VSOCK_EVENTS_COUNT);
+
+            let vsock_box = Box::new(
+                devices::virtio::Vsock::new(u64::from(cfg.guest_cid), epoll_config, backend)
+                    .map_err(StartMicrovmError::CreateVsockDevice)?,
+            );
+
+            builder.attach_device(
+                cfg.vsock_id.clone(),
+                MmioDevice::new(builder.inner().guest_memory().clone(), vsock_box).map_err(
+                    |e| {
+                        StartMicrovmError::RegisterMMIODevice(
+                            super::device_manager::mmio::Error::CreateMmioDevice(e),
+                        )
+                    },
+                )?,
+            )?;
+        }
+
+        Ok(())
+    }
+
+    fn init_guest_memory(&mut self) -> std::result::Result<(), StartMicrovmError> {
+        if self.guest_memory.is_none() {
```

Comment:
It is redundant to keep both guest_memory and vmm around in the vmm_controller.
Why not just have a function for returning `arch_mem_regions` and then in `start_microvm` you do:

```
fn arch_mem_regions() -> Vec<(GuestAddress, usize)>{
    let mem_size = self
                .vm_config
                .mem_size_mib
                .ok_or(StartMicrovmError::GuestMemory(
                    memory_model::GuestMemoryError::MemoryNotInitialized,
                ))?
                << 20;
            arch::arch_memory_regions(mem_size)
}

fn start_microvm() {
    let arch_mem_regions = self.arch_mem_regions();
    let builder_config = VmmBuilderConfig {
            guest_memory: GuestMemory::new(&arch_mem_regions).map_err(StartMicrovmError::GuestMemory)?),
            kernel_entry_addr,
            kernel_cmdline: kernel_config.cmdline,
            vcpu_config,
            seccomp_level: self.seccomp_level,
        };
}
```